### PR TITLE
core(plugins): allow restricted gatherers

### PR DESCRIPTION
**Pull request by connorjclark**
I jumped on this b/c proxies are cool and I wanted to learn how the config parsing /  plugins work.

draft b/c @paulirish is still exploring / writing a doc for plugin gatherers. will undraft if we figure this approach is 💯.

### CODE REVIEW

File: lighthouse-core/gather/gather-runner.js [hunk 1]
```diff
@@ -396,8 +435,12 @@ class GatherRunner {
 
       // Add gatherer options to the passContext.
       passContext.options = gathererDefn.options || {};
+      const gathererPassContext = this._makeGathererPassContext(passContext, {
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/gather/gather-runner.js]: oh no sorry I mean the opposite, move the `gathererDefn.options ?` into `_makeGathererPassContext` :)

[the remark this answers is not in the document]
Comment (connorjclark) [reply lighthouse-core/gather/gather-runner.js]: Oh, right. I figured it was a bit _weird_ that it relies on `passContext.options` being set in that way. But we already do it, so guess we might as well run with it.

**Comment by brendankenny** [comment]
I was imagining something far more limited, e.g. provide a single function to run in the page and the output will be the plugin artifact.

It would be good to start with an actual use case to motivate what's needed and be able to give context to design tradeoffs (e.g. if we're providing a pass context we're committing to that form of pass context, etc). Are we talking collecting all iframes and checking if react is in dev mode or something more?

**Comment by connorjclark** [comment]
> I was imagining something far more limited, e.g. provide a single function to run in the page and the output will be the plugin artifact.

That's fair. And matches one of the already-existing ways to define a gatherer in the config. I'll point out though, that this restricted Gatherer approach (JS Proxy class or otherwise) is required if we want to expand a bit beyond just `evaluateAsync` in the future.

> It would be good to start with an actual use case to motivate what's needed and be able to give context to design tradeoffs (e.g. if we're providing a pass context we're committing to that form of pass context, etc). Are we talking collecting all iframes and checking if react is in dev mode or something more?

At least for the react dev mode case, `beforePass` logic is necessary (see #9511. a hook must be installed before React loads). So, that precludes the simple "provide one function to be run in the `afterPass`" approach.

**Comment by connorjclark** [comment]
Currently aren't focused on plugins, and maybe won't go with this PR's approach, so closing for now.

**Pull request closed without merging**